File: CMP326Fall2019MidtermProject1/gameinterfaces/Creature.java
package gameinterfaces;

/**
 * 
 * A class that implements this interface should have the following attributes:
 * <ul>
 * <li>name: String</li>
 * 
 * <li>description: String</li>
 * 
 * <li>hitPoints: int</li>
 * 
 * <li>damage: int</li>
 * 
 * <li>room: Room</li>
 * </ul>
 * 
 * The implementing class should only have the following constructor:
 * 
 * public Creature(String name, String description, int hitPoints, int damage)
 * 
 * room should be initialized to null 
 * 
 * provide getters for the following attributes: name, description, hitPoints, damage, Room.
 * 
 * provide setters for the following attributes: hitPoints, damage, Room.
 * 
 * 
 */

public interface Creature {

	/**
	 * 
	 * This method displays a message containing the following:
	 * <br>
	 * "<code>this.name</code> attacks <code>target creature's name</code>, dealing <code>damage</code> damage"
	 * 
	 * @param creature A reference to the target Creature object being attacked
	 * 
	 */
	public void attack(Creature creature);

	/**
	 * 
	 * This method subtracts the specified amount of <code>damage</code> from <code>hitPoits</code>. Please note that
	 * <code>hitPoints</code> cannot go below zero.
	 * 
	 * @param damage The amount of damage inflicted on <code>this</code> Creature.
	 * 
	 */
	public void takeDamage(int damage);

	
	/**
	 * This method return <code>true</code> if   <code>this</code> Creature has positive <code>hitPoints</code>.
	 * 
	 * @return <code>true</code> if   <code>this</code> Creature has positive <code>hitPoints</code>, <code>false</code> otherwise.
	 * 
	 */
	public boolean isAlive();

	/**
	 * 
	 * @return A String formatted as follows:
	 * <br>
	 * <br>
	 * "name %15s, description %20s, hitPoints %03d, damage %02d" 
	 */
	@Override
	public String toString();

}

File: Qt_workspace/rpg_txt_game/creature.cpp
#include "creature.h"

Creature::Creature(){
    this->name = "";
    this->description = "";
    this->hitPoints = 0;
    this->damage = 0;
    this->room = Room();
}

Creature::Creature(string name, string description,
                   int hitPoints, int damage)
{
    this->name = name;
    this->description = description;
    this->hitPoints = hitPoints;
    this->damage = damage;
    this->room = Room();
}

void Creature::attack(Creature creature){
    cout << this->getName() << " attacks " + creature.getName() << ", dealing "
         << this->getDamage() << " damage(s)" <<endl;
}

void Creature::takeDamage(int damage){
    this->setDamage(this->getDamage()-damage);
}

bool Creature::isAlive(){
    return this->getHitPoint() > 0 ? true : false;
}

string Creature::toString(){
    return "name "+ this->getName() + "description"+this->getDescription() +
            "hitPoints" + to_string(this->getHitPoint()) + "damage "+ to_string(this->getDamage());
}

void Creature::setHitPoints(int hitPoints){
    this->hitPoints = hitPoints < 0 ? 0 : hitPoints;
}

string Creature::getName(){
    return this->name;
}

string Creature::getDescription(){
    return this->description;
}

int Creature::getHitPoint(){
    return this->hitPoints;
}

int Creature::getDamage(){
    return this->damage;
}

Room Creature::getRoom(){
    return this->room;
}
